content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/vmware_best_fit_least_utilized.rb

```diff
@@ -1,37 +1,74 @@
-ems = vm.ext_management_system
-raise "EMS not found for VM [#{vm.name}]" if ems.nil?
-
-# Log space required
-$evm.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
-
-host = storage = nil
-min_registered_vms = nil
-prov.eligible_hosts.each do |h|
-  next if h.maintenance
-  next unless h.power_state == "on"
-  nvms = h.vms.length
-  if min_registered_vms.nil? || nvms < min_registered_vms
-    storages = h.writable_storages.find_all { |s| s.free_space > vm.provisioned_storage } # Filter out storages that do not have enough free space for the Vm
-    s = storages.max_by(&:free_space)
-    unless s.nil?
-      host    = h
-      storage = s
-      min_registered_vms = nvms
+module ManageIQ
+  module Automate
+    module Infrastructure
+      module VM
+        module Provisioning
+          module Placement
+            class VmwareBestFitLeastUtilized
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def main
+                @handle.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
+                best_fit_least_utilized
+              end
+
+              private
+
+              def request
+                @request ||= @handle.root["miq_provision"].tap do |req|
+                  log_and_raise('miq_provision object not passed in') if req.nil?
+                end
+              end
+
+              def vm
+                @vm ||= request.vm_template.tap do |vm|
+                  log_and_raise('VM not specified') if vm.nil?
+                end
+              end
+
+              def storage_profile_id
+                @storage_profile_id ||= request.get_option(:placement_storage_profile).tap do |sp|
+                  @handle.log("info", "Selected storage_profile_id: #{sp}") if sp
+                end
+              end
+
+              def best_fit_least_utilized
+                host = storage = min_registered_vms = nil
+                request.eligible_hosts.select { |h| !h.maintenance && h.power_state == "on" }.each do |h|
+                  next if min_registered_vms && h.vms.size >= min_registered_vms
+
+                  storages = h.writable_storages.find_all { |s| s.free_space > vm.provisioned_storage } # Filter out storages that do not have enough free space for the Vm
+                  storages.select! { |s| s.storage_profiles.pluck(:id).include?(storage_profile_id) } if storage_profile_id
+
+                  s = storages.max_by(&:free_space)
+                  next if s.nil?
+                  host    = h
+                  storage = s
+                  min_registered_vms = h.vms.size
+                end
+
+                # Set host and storage
+                request.set_host(host) if host
+                request.set_storage(storage) if storage
+
+                @handle.log("info", "vm=[#{vm.name}] host=[#{host}] storage=[#{storage}]")
+              end
+
+              def log_and_raise(message)
```

Comment:
@lfu 
Can you replace this method with this one
https://github.com/ManageIQ/manageiq-content/blob/a50ea6c95dfb2e2b51a69720e67f0fae973d1733/content/automate/ManageIQ/System/CommonMethods/Utils.class/__methods__/log_object.rb#L56

In the CFME Explorer you would have to add an embedded method pointing to this method and replace all calls to log_and_raise to use
```
ManageIQ::Automate::System::CommonMethods::Utils::LogObject.log_and_raise(message, @handle)
```

To help with the spec and embed methods look at this PR
https://github.com/ManageIQ/manageiq-content/pull/251/files

# This is the embedded method link, its only needed in the spec
 # Not in the actual file and ensures that the symbols are satisfied
 require File.join(ManageIQ::Content::Engine.root, 'content/automate/ManageIQ/System/CommonMethods/Utils.class/__methods__/log_object.rb')






       # This is optional so that we dont have to the run the embedded method spec
       # This is a stub and returns canned responses with no computational overhead
       #
       allow(ManageIQ::Automate::System::CommonMethods::Utils::LogObject).to receive(:log_and_raise).with(msg, ae_service)